File: AWS/iac-microbenchmark/ec2/go/sha256/main.go
// main.go
package main

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"os"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/kms"
	"github.com/aws/aws-sdk-go-v2/service/kms/types"
)

const SIGN_ALGORITHM = "HMAC_SHA_256"

func main() {
	// Get the JSON input from command line argument
	requestJSONRaw := os.Args[1]

	var requestJSON map[string]string
	if err := json.Unmarshal([]byte(requestJSONRaw), &requestJSON); err != nil {
		fmt.Printf("Error unmarshalling JSON: %v\n", err)
		return
	}

	message, ok := requestJSON["message"]
	if !ok {
		fmt.Println("Error: message field is required")
		return
	}

	shaKmsKeyID := os.Getenv("SHA256_KMS_KEY_ARN")
	if shaKmsKeyID == "" {
		fmt.Println("Error: SHA256_KMS_KEY_ARN environment variable is not set")
		return
	}

	// Load the default AWS configuration
	cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion("us-east-1"))
	if err != nil {
		fmt.Printf("unable to load SDK config, %v\n", err)
		return
	}

	// Create a new KMS client
	kmsClient := kms.NewFromConfig(cfg)

	// Convert the message to bytes
	messageBytes := []byte(message)

	// Use KMS to generate HMAC for the message
	input := &kms.GenerateMacInput{
		KeyId:        aws.String(shaKmsKeyID),
		Message:      messageBytes,
		MacAlgorithm: types.MacAlgorithmSpecHmacSha256,
	}

	response, err := kmsClient.GenerateMac(context.TODO(), input)
	if err != nil {
		fmt.Printf("Error generating MAC: %v\n", err)
		return
	}

	// Base64 encode the signature for output
	signature := base64.StdEncoding.EncodeToString(response.Mac)

	// Prepare the result
	resultDict := map[string]string{
		"signature": signature,
	}

	resultJSON, err := json.Marshal(resultDict)
	if err != nil {
		fmt.Printf("Error marshalling result to JSON: %v\n", err)
		return
	}

	fmt.Println(string(resultJSON))
}

File: AWS/iac-microbenchmark/ec2/rust/sha256/src/main.rs
use aws_sdk_kms::{self as kms, primitives::Blob};
use base64::encode;
use serde::{Deserialize, Serialize};
use std::env;
use std::error::Error;

#[derive(Serialize, Deserialize, Debug)]
struct MessageStruct {
    message: String,
}

#[derive(Serialize, Deserialize, Debug)]
struct SignatureStruct {
    signature: String,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    // Get the KMS key ARN from environment variables
    let sha_kms_key_id = env::var("SHA256_KMS_KEY_ARN").map_err(|_| "KMS key ARN not set")?;

    // Get JSON input from command line
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        return Err("Usage: ./rustexecutable '{\"message\":\"your_message_here\"}'".into());
    }

    let body_str = &args[1];

    // Parse the input JSON
    let message_struct: MessageStruct = serde_json::from_str(body_str)
        .map_err(|_| "Invalid Request Body, missing message or incorrect format.")?;

    // Convert the message to bytes
    let message_blob = Blob::new(message_struct.message.as_bytes());

    // Initialize the KMS client
    let config = aws_config::load_from_env().await;
    let kms_client = kms::Client::new(&config);

    // Attempt to sign the message using KMS
    match kms_sign_message(&kms_client, &sha_kms_key_id, message_blob).await {
        Ok(signature) => {
            let sig_struct = SignatureStruct { signature };

            // Serialize the struct to a JSON string
            let str_sig_struct = serde_json::to_string(&sig_struct)?;
            println!("{}", str_sig_struct);

            Ok(())
        }
        Err(e) => Err(format!("Error signing message: {}", e).into()),
    }
}

/// Function to sign a message using AWS KMS.
async fn kms_sign_message(
    kms_client: &kms::Client,
    key_id: &str,
    message: Blob,
) -> Result<String, Box<dyn Error>> {
    let mac_output = kms_client
        .generate_mac()
        .key_id(key_id)
        .message(message)
        .mac_algorithm(aws_sdk_kms::types::MacAlgorithmSpec::HmacSha256)
        .send()
        .await
        .map_err(|err| format!("Failed to generate mac hash: {}", err))?;

    let mac = mac_output.mac.ok_or_else(|| "No Mac hash returned")?;

    Ok(encode(mac.as_ref()))
}
